File: src/Network/Endpoint.cpp
/*
** EPITECH PROJECT, 2026
** myftp
** File description:
** Endpoint
*/

#include "Endpoint.hpp"

#include <array>
#include <iostream>
#include <stdexcept>
#include <arpa/inet.h>

#include "constants.hpp"

namespace network {
Endpoint::Endpoint(const int &port, const std::string &hostName): _port{port},
    _hostName{hostName}
{
    _address.sin_family = AF_INET;
    _address.sin_port   = htons(_port);
    if (hostName.empty())
        _address.sin_addr.s_addr = INADDR_ANY;
    else {
        const int inetResult = inet_pton(AF_INET, hostName.c_str(),
            &_address.sin_addr);

        if (inetResult == 0)
            throw std::runtime_error("Invalid network address");
        if (inetResult == -1)
            throw std::runtime_error("Invalid network address family");
    }
}

Endpoint::Endpoint(const sockaddr_in &address): _address{address}
{}

int Endpoint::getPort() const noexcept
{
    return ntohs(_address.sin_port);
}

std::string Endpoint::getHostname() const noexcept
{
    std::array<char, INET_ADDRSTRLEN> buf{};
    const char *res = inet_ntop(
        AF_INET,
        &_address.sin_addr,
        buf.data(),
        buf.size()
        );
    if (!res) {
        return {};
    }
    return std::string{buf.data()};
}

sockaddr_in &Endpoint::getAddress() noexcept
{
    return _address;
}
} // ftp

File: src/AIClient/network/src/network/endpoint.rs
use libc::in_addr_t;
use std::net::Ipv4Addr;

pub struct Endpoint {
    port: u16,
    hostname: String,
    address: libc::sockaddr_in,
}

impl Endpoint {
    pub fn new(port: u16, host: &str) -> Self {
        Endpoint {
            port,
            hostname: host.to_string(),
            address: libc::sockaddr_in {
                sin_family: libc::AF_INET as libc::sa_family_t,
                sin_port: libc::htons(port),
                sin_addr: libc::in_addr {
                    s_addr: match host.is_empty() {
                        true => libc::INADDR_ANY as in_addr_t,
                        false => u32::from(host.parse::<Ipv4Addr>().unwrap()).to_be(),
                    },
                },
                sin_zero: [0; 8],
            },
        }
    }

    pub fn from_sockaddr(address: libc::sockaddr_in) -> Self {
        Endpoint {
            port: libc::ntohs(address.sin_port),
            hostname: Ipv4Addr::from(u32::from_be(address.sin_addr.s_addr)).to_string(),
            address,
        }
    }

    pub fn port(&self) -> u16 {
        self.port
    }

    pub fn host(&self) -> &str {
        &self.hostname
    }

    pub fn address(&self) -> libc::sockaddr_in {
        self.address
    }
}
